Mathlib/LinearAlgebra/LinearDisjoint.lean

```diff
@@ -0,0 +1,550 @@
+  rw [← hx', ← hy']; congr
+  exact (H N' hN).injective (by simp [← mulMap_comp_lTensor _ hN, hx', hy', hxy])
+
+theorem of_linearDisjoint_finite
+    (H : ∀ (M' N' : Submodule R S), M' ≤ M → N' ≤ N →
+      [Module.Finite R M'] → [Module.Finite R N'] → M'.LinearDisjoint N') :
+    M.LinearDisjoint N :=
+  of_linearDisjoint_finite_left _ _ fun _ hM _ ↦
+    of_linearDisjoint_finite_right _ _ fun _ hN _ ↦ H _ _ hM hN
+
+end LinearDisjoint
+
+end Semiring
+
+section CommSemiring
+
+variable [CommSemiring R] [CommSemiring S] [Algebra R S]
+
+variable {M N : Submodule R S}
+
+/-- Linearly disjoint is symmetric in a commutative ring. -/
+theorem LinearDisjoint.symm (H : M.LinearDisjoint N) : N.LinearDisjoint M :=
+  H.symm_of_commute fun _ _ ↦ mul_comm _ _
+
+/-- Linearly disjoint is symmetric in a commutative ring. -/
+theorem linearDisjoint_symm : M.LinearDisjoint N ↔ N.LinearDisjoint M :=
+  ⟨LinearDisjoint.symm, LinearDisjoint.symm⟩
+
+end CommSemiring
+
+section Ring
+
+namespace LinearDisjoint
+
+variable [CommRing R] [Ring S] [Algebra R S]
+
+variable (M N : Submodule R S)
+
+variable {M N} in
+theorem map_linearIndependent_left_of_flat (H : M.LinearDisjoint N) [Module.Flat R N]
+    {ι : Type*} {m : ι → M} (hm : LinearIndependent R m) : LinearMap.ker (mulLeftMap N m) = ⊥ := by
+  refine LinearMap.ker_eq_bot_of_injective ?_
+  simp_rw [mulLeftMap_eq_mulMap_comp, LinearMap.coe_comp, LinearEquiv.coe_coe,
+    ← Function.comp.assoc, EquivLike.injective_comp]
+  rw [LinearIndependent, LinearMap.ker_eq_bot] at hm
+  exact H.injective.comp (Module.Flat.rTensor_preserves_injective_linearMap (M := N) _ hm)
+
+theorem of_map_linearIndependent_left {ι : Type*} (m : Basis ι R M)
+    (H : LinearMap.ker (mulLeftMap N m) = ⊥) : M.LinearDisjoint N := by
+  -- need this instance otherwise `LinearMap.ker_eq_bot` does not work
+  letI : AddCommGroup (ι →₀ N) := Finsupp.instAddCommGroup
+  exact of_map_linearIndependent_left' M N m (LinearMap.ker_eq_bot.1 H)
+
+variable {M N} in
+theorem map_linearIndependent_right_of_flat (H : M.LinearDisjoint N) [Module.Flat R M]
+    {ι : Type*} {n : ι → N} (hn : LinearIndependent R n) : LinearMap.ker (mulRightMap M n) = ⊥ := by
+  refine LinearMap.ker_eq_bot_of_injective ?_
+  simp_rw [mulRightMap_eq_mulMap_comp, LinearMap.coe_comp, LinearEquiv.coe_coe,
+    ← Function.comp.assoc, EquivLike.injective_comp]
+  rw [LinearIndependent, LinearMap.ker_eq_bot] at hn
+  exact H.injective.comp (Module.Flat.lTensor_preserves_injective_linearMap (M := M) _ hn)
+
+theorem of_map_linearIndependent_right {ι : Type*} (n : Basis ι R N)
+    (H : LinearMap.ker (mulRightMap M n) = ⊥) : M.LinearDisjoint N := by
+  -- need this instance otherwise `LinearMap.ker_eq_bot` does not work
+  letI : AddCommGroup (ι →₀ M) := Finsupp.instAddCommGroup
+  exact of_map_linearIndependent_right' M N n (LinearMap.ker_eq_bot.1 H)
+
+variable {M N} in
+theorem map_linearIndependent_mul_of_flat_left (H : M.LinearDisjoint N) [Module.Flat R M]
```

Comment:
What is the `map` referring to here?

Reply:
It means that it maps two linearly independent families to a new linear independent family, which is the product of these two families. It's written in the module docstring in the beginning of the file:

>  if `M` and `N` are linearly disjoint, if one of `M` and `N` is flat, then for any family of
>  `R`-linearly independent elements `{ m_i }` of `M`, and any family of
>  `R`-linearly independent elements `{ n_j }` of `N`, the family `{ m_i * n_j }` in `S` is
>  also `R`-linearly independent.

I guess I should also copy the docstring to each function.

Reply:
Docstring added.

Reply:
I see. Do you think it's a good idea to remove `map_` from the name of other results?